docs/melange-book/celsius-converter-exception.md

```diff
@@ -0,0 +1,333 @@
+# Celsius Converter
+
+Our second widget will be one that takes a temperature value in Celsius and
+converts it to Fahrenheit. Create a new file in `src` called
+`CelsiusConverter.re`:
+
+```reasonml
+let convert = celsius => 9.0 /. 5.0 *. celsius +. 32.0;
+
+[@react.component]
+let make = () => {
+  let (celsius, setCelsius) = React.useState(() => "");
+
+  <div>
+    <input
+      value=celsius
+      onChange={evt => {
+        let newCelsius = ReactEvent.Form.target(evt)##value;
+        setCelsius(_ => newCelsius);
+      }}
+    />
+    {React.string({js|°C = |js})}
+    {React.string({js|? °F|js})}
+  </div>;
+};
+```
+
+Inside the `input`'s `onChange` handler, we get the event target using
+`ReactEvent.Form.target`, which has the type `ReactEvent.Form.t => {_.. }`. What
+is `{_.. }`? It's shorthand for a [Js.t
+object](../communicate-with-javascript.md#using-jst-objects), a polymorphic
+object whose fields aren't well-defined. In OCaml, we use the `##` operator to
+access fields on an object. Be careful when using objects, as it subverts
+OCaml's type system. We could write `ReactEvent.Form.target(evt)##value + 1`,
+treating it as if it were an integer, and the compiler wouldn't complain.
```

Comment:
I think this section needs some more details. I am not sure how far you want to take it, but I think at least
- the signature should be reviewed
- i am not sure about the "as it subverts OCaml's type system", because it suggests the compiler behaves in unexpected ways, but the behavior is quite clear and documented.

Some details:

The type of `ReactEvent.Form.target` is `ReactEvent.Form.t => Js.t({..})` (not sure where the `{_ ..}` came from? I have never seen that syntax).

This type is composed of two parts:
- The `Js.t` represents the type of JavaScript objects in Melange
- `{..}` is the type of OCaml objects that does not add any constraint in the methods or values they contain

So `Js.t({..})` is the type of essentially any JavaScript object.

If the type was `Js.t({..size: int})` it would be the type of all JavaScript objects that contain at least a `size` property of type `int`.

Maybe we could include some example:

```reason
let t: Js.t({..}) = {"foo": 2} /* Can be object with property foo */
let t: Js.t({..}) = {"bar": 2} /* Can be object with property bar */
let t: Js.t({..size: int}) = {"bar": 2} /* Can't be object without property size */
```

Reply:
Added more clarification of Js.t in https://github.com/melange-re/melange-re.github.io/pull/93/commits/0dfdf62d19e2d78beb2e40ed44045e2bab8633d9